### PR TITLE
NERDTree keeps 'No Name' buffer after deleting current node 'md' #755 (closes #755)

### LINKED ISSUE #755: NERDTree keeps 'No Name' buffer after deleting current node 'md'
_To assist in resolving your issue, provide as much information as possible, in place of the ellipses (`…`) below._

---
**Environment:** _Describe your Vim/NERDTree setup._

>* Operating System: Archlinux, Linux Opensuse and Window 7
>* GVim version `:version`: 8.0 Included patch 1-69
>* Vim version `:version`: 8.0 Included patch 1-2
>* NERDTree version `git rev-parse --short HEAD`: f554c20
>* NERDTree settings applied in your vimrc, if any:
let NERDTreeWinPos = "left"
let NERDTreeWinSize = 35
let NERDTreeIgnore = ['\.job$', '^CVS$', '\.orig', '\.log$']
let g:NERDTreeDirArrows = 0
let g:NERDTreeStatusline = "%f"
let g:NERDTreeDirArrowExpandable = ''
let g:NERDTreeDirArrowCollapsible = ''

**Process:** _List the steps that will recreate the issue._

>0. Open several buffers from the tree
>1. Delete one file from the tree using md
   NERDTree ask for confirmation to delete the file and delete the associated buffer if open
>2. confirm delete file and buffer with 'y'

**Current Result:** _Describe what you you currently experience from this process._

> Buffer is actually deleted but it remains open as 'No Name'

To completely remove residue buffer one should call ':Bdelete'

/!\ I tested (re) this several time before submitting the issue and notice that sometime:
- residue buffer is replaced with newly open file
- If deleted buffer was the first one to be opened it will be removed completely without leaving any residue buffer
- **This is occurring only if the file to be deleted is the currently active buffer**

---
**Optional**

**Screenshot(s):**

image

**Possible Fix:** _(Have you poked around in the code?)_
The setting below doesn't affect the results.
```vim
"Automatically delete the buffer after deleting or renaming a file
if !exists("g:NERDTreeAutoDeleteBuffer")
  let g:NERDTreeAutoDeleteBuffer = 0
endif
```
There is one function to be reviewed
```vim
function! s:promptToDelBuffer(bufnum, msg)
```

I think the issue will be solved by changin the line below in **s:promptToDelBuffer**:
- from
```vim
exec "tabdo windo if winbufnr(0) == " . a:bufnum . " | exec ':enew! ' | endif"
```
- to
```vim
exec "tabdo windo if winbufnr(0) == " . a:bufnum . " | exec ':bnext! ' | endif"
```

### PR DESCRIPTION
**Pull request by mboughaba**
Go to the new buffer in buffer list instead of creating a new buffer.
If only one buffer is open, now we go back to nerdtree buffer.

### CODE REVIEW

**Comment by lifecrisis** [comment]
I'm leaning toward accepting this change. It avoids creating a new buffer, which I like.

@PhilRunninger, do you approve?

**Comment by PhilRunninger** [comment]
This changes the current behavior when only one non-NERDTree buffer is open. It causes NERDTree to be the only window after `md`. This leads to window sizing issues that have caused a lot of confusion for other users. Maybe it would be better to leave the old behavior if deleting the last `listed` buffer. But I do like this change in all other situations.

**Comment by mboughaba** [comment]
>This changes the current behavior when only one non-NERDTree buffer is open. It causes NERDTree to be the only window after md. This leads to window sizing issues that have caused a lot of confusion for other users. Maybe it would be better to leave the old behavior if deleting the last listed buffer. But I do like this change in all other situations.

Thanks guys! I agree.
I tried at least to put that in small words in my commit

> If only one buffer is open, now we go back to nerdtree buffer.

I'll work on that

File: nerdtree_plugin/fs_menu.vim [hunk 1]
```diff
@@ -55,7 +55,14 @@ function! s:promptToDelBuffer(bufnum, msg)
         " Is not it better to close single tabs with this file only ?
         let s:originalTabNumber = tabpagenr()
         let s:originalWindowNumber = winnr()
-        exec "tabdo windo if winbufnr(0) == " . a:bufnum . " | exec ':enew! ' | endif"
+        let s:listedBufferCount = len(getbufinfo({'buflisted':1}))
```
Comment (lifecrisis) [inline nerdtree_plugin/fs_menu.vim]: Note that `getbufinfo()` is not supported in Vim 7.4.

Comment (lifecrisis) [reply nerdtree_plugin/fs_menu.vim]: I can't approve the change until this is fixed.

Comment (mboughaba) [reply nerdtree_plugin/fs_menu.vim]: Thanks for the find. I wasn't aware that getbufinfo is not supported in 7.4. Will work on that.

Comment (mboughaba) [reply nerdtree_plugin/fs_menu.vim]: Below works on vim 7.2 in my linux machine
```vim
:echo len(filter(range(1, bufnr('$')), 'buflisted(v:val)'))
```
I Need to find a clean way to do this

Comment (PhilRunninger) [reply nerdtree_plugin/fs_menu.vim]: I'm fine with it either way - old syntax or both.

Comment (lifecrisis) [reply nerdtree_plugin/fs_menu.vim]: @mboughaba, I'll eventually push for a fork that will break with Vim 7 and only support Vim 8. No point in doing it right now though, adoption is not wide enough to to warrant the move.

Believe me, if we break compatibility with 7 at this time, complaints will flow!

**Pull request merged**